src/canard.c

```diff
@@ -1,7 +1,919 @@
+    CanardRxState* state = (CanardRxState*)canardAllocateBlock(allocator);
+
+    if (state == NULL)
+    {
+        return NULL;
+    }
+    memcpy(state, &init, sizeof *state);
+
+    return state;
+}
+
+/**
+ * This function can be invoked by the application to release pool blocks that are used
+ * to store the payload of this transfer
+ */
+CANARD_INTERNAL uint64_t canardReleaseStatePayload(CanardInstance* ins, CanardRxState* rxstate)
+{
+    CanardBufferBlock* temp = rxstate->buffer_blocks;
+    while (rxstate->buffer_blocks != NULL)
+    {
+        temp = rxstate->buffer_blocks->next;
+        canardFreeBlock(&ins->allocator, rxstate->buffer_blocks);
+        rxstate->buffer_blocks = temp;
+    }
+    rxstate->payload_len = 0;
+    return 0;
+}
+
+/**
+ *  CanardBufferBlock functions
+ */
+
+/**
+ * pushes data into the rx state. Fills the buffer head, then appends data to buffer blocks
+ */
+CANARD_INTERNAL void canardBufferBlockPushBytes(CanardPoolAllocator* allocator, CanardRxState* state,
+                                                const uint8_t* data,
+                                                uint8_t data_len)
+{
+    uint16_t data_index = 0;
+    uint16_t i;
+
+    // if head is not full, add data to head
+    if ((int)CANARD_RX_PAYLOAD_HEAD_SIZE - (int)state->payload_len > 0)
+    {
+        for (i = state->payload_len; i<CANARD_RX_PAYLOAD_HEAD_SIZE && data_index<data_len; i++, data_index++)
+        {
+            state->buffer_head[i] = data[data_index];
+        }
+        if (data_index >= data_len /*- 1*/)
+        {
+            state->payload_len += data_len;
+            return;
+        }
+    } // head is full.
+
+    uint8_t num_buffer_blocks =
+        (((state->payload_len + data_len) - CANARD_RX_PAYLOAD_HEAD_SIZE) / CANARD_BUFFER_BLOCK_DATA_SIZE) + 1;
+    uint8_t index_at_nth_block = (((state->payload_len) - CANARD_RX_PAYLOAD_HEAD_SIZE) % CANARD_BUFFER_BLOCK_DATA_SIZE);
+
+    // get to current block
+    CanardBufferBlock* block;
+    uint8_t nth_block = 1;
+
+    // buffer blocks uninitialized
+    if (state->buffer_blocks == NULL)
+    {
+        state->buffer_blocks = canardCreateBufferBlock(allocator);
+        block = state->buffer_blocks;
+        index_at_nth_block = 0;
+    }
+    else
+    {
+        // get to block
+        block = state->buffer_blocks;
+        while (block->next != NULL)
+        {
+            nth_block++;
+            block = block->next;
+        }
+        if (num_buffer_blocks > nth_block && index_at_nth_block == 0)
+        {
+            block->next = canardCreateBufferBlock(allocator);
+            block = block->next;
+            nth_block++;
+        }
+    }
+
+    // add data to current block until it becomes full, add new block if necessary
+    while (data_index < data_len)
+    {
+        for (i = index_at_nth_block; i<CANARD_BUFFER_BLOCK_DATA_SIZE && data_index<data_len; i++, data_index++)
+        {
+            block->data[i] = data[data_index];
```

Comment:
`block` may be NULL if the pool is exhausted. This will lead to null pointer dereferencing here.

We need to return some status code from this function and handle it in the caller. If new data could not be pushed we probably need to abort reception of the transfer.